### PR TITLE
Amgcl mpi

### PR DESCRIPTION
**Pull request by RiccardoRossi**
as the title suggests, this exposes full AMG capabilities in MPI (on the basis of the AMGCL lib of @ddemidov )

### CODE REVIEW

**Comment by philbucher** [comment]
no test?

**Comment by RiccardoRossi** [comment]
test_trilinos_linear_solvers.py tests this (and passes)

**Comment by RiccardoRossi** [comment]
expanding about the tests, this substitutes the old version of the AMGCL_MPI_Linear solver, so the test remains the same

File: kratos/linear_solvers/amgcl_solver.h [hunk 1]
```diff
@@ -101,25 +101,29 @@ AMGCLScalarSolve(
...
 
+    // amgcl::make_solver<
+    //     amgcl::amg<
+    //         Backend,
+    //         amgcl::runtime::coarsening::wrapper,
+    //         amgcl::runtime::relaxation::wrapper
+    //         >,
+    //     amgcl::runtime::solver::wrapper<Backend>
+    //     > solve(amgcl::adapter::zero_copy(TSparseSpaceType::Size1(rA), rA.index1_data().begin(), rA.index2_data().begin(), rA.value_data().begin()), amgclParams);
+
```
Comment (ddemidov) [inline kratos/linear_solvers/amgcl_solver.h]: Personally, I would just remove the commented lines above. Git can track your changes for you.

File: kratos/linear_solvers/amgcl_solver.h [hunk 2]
```diff
@@ -149,19 +154,20 @@ AMGCLBlockSolve(
...
 
+    // amgcl::make_solver<
+    //     amgcl::amg<
+    //         Backend,
+    //         amgcl::runtime::coarsening::wrapper,
+    //         amgcl::runtime::relaxation::wrapper
+    //         >,
+    //     amgcl::runtime::solver::wrapper<Backend>
+    //     > solve( amgcl::adapter::block_matrix<value_type>(std::tie(n,rA.index1_data(),rA.index2_data(),rA.value_data() )), amgclParams);
+
```
Comment (ddemidov) [inline kratos/linear_solvers/amgcl_solver.h]: Same here

File: kratos/linear_solvers/amgcl_solver.h [hunk 3]
```diff
@@ -115,7 +115,7 @@ AMGCLScalarSolve(
     std::tie(rIterationNumber, rResidual) = solve(rB, rX);
 
     if(verbosity_level > 1 )
-        std::cout << "AMGCL Memory Occupation (MB): " << amgcl::backend::bytes(solve) << std::endl;
+        std::cout << "AMGCL Memory Occupation (kB): " << amgcl::backend::bytes(solve)/1024 << std::endl;
```
Comment (ddemidov) [inline kratos/linear_solvers/amgcl_solver.h]: There is also `amgcl::human_readable_memory()` available which will give you the most appropriate units with 2 digits precision:

https://github.com/ddemidov/amgcl/blob/d1f8387a05ccc409d57505460d5fde27637ce902/amgcl/util.hpp#L310-L320

**Comment by RiccardoRossi** [comment]
@ddemidov you fine with the current version? if so i would merge...

**Review by ddemidov — APPROVED** [review]

**Review by loumalouomega — APPROVED** [review]

**Comment by RiccardoRossi** [comment]
thx @loumalouomega

**Comment by loumalouomega** [comment]
imagen

**Comment by RiccardoRossi** [comment]
well...and of course thx to @ddemidov who did most of the job :-p

**Pull request merged**